### PR TITLE
[IMP] purchase_stock: price diff as val. layer

### PR DESCRIPTION
**Pull request by svs-odoo**
Removes the price difference account field on product template and product category, and instead of using this account, in case of price difference between the purchase order and the bill, creates a stock valuation layer about this price diff.

task-2081607

### CODE REVIEW

File: addons/purchase_stock/models/account_invoice.py [hunk 1]
```diff
@@ -1,13 +1,19 @@
...
 
 
 class AccountMove(models.Model):
     _inherit = 'account.move'
 
+    def button_draft(self):
+        res = super().button_draft()
+        # Deletes already created stock valuation layers.
+        self.stock_valuation_layer_ids.sudo().unlink()
+        return res
```
Comment (sle-odoo) [inline addons/purchase_stock/models/account_invoice.py]: unlinking posted move?? humm..
this override should be in stock_account

File: addons/stock_account/models/account_move.py [hunk 2]
```diff
@@ -93,7 +93,7 @@ def _stock_account_prepare_anglo_saxon_out_lines_vals(self):
         '''
         lines_vals_list = []
         for move in self:
-            if not move.is_sale_document(include_receipts=True) or not move.company_id.anglo_saxon_accounting:
+            if (not move.is_sale_document(include_receipts=True) or not move.company_id.anglo_saxon_accounting) and not already_out:
```
Comment (sle-odoo) [inline addons/stock_account/models/account_move.py]: seems weird to bypass this logic when already out, i think we should add more accounting entries if the product is already sold but. can you explain?

File: addons/stock_account/models/stock_move.py [hunk 3]
```diff
@@ -406,29 +406,7 @@ def _generate_valuation_lines_data(self, partner_id, qty, debit_value, credit_va
...
-                'quantity': qty,
-                'product_uom_id': self.product_id.uom_id.id,
-                'ref': description,
-                'partner_id': partner_id,
-                'credit': diff_amount > 0 and diff_amount or 0,
-                'debit': diff_amount < 0 and -diff_amount or 0,
-                'account_id': price_diff_account.id,
-            }
-        return rslt
+        return {'credit_line_vals': credit_line_vals, 'debit_line_vals': debit_line_vals}
```
Comment (sle-odoo) [inline addons/stock_account/models/stock_move.py]: 👍

File: addons/purchase_stock/models/account_invoice.py [hunk 4]
```diff
@@ -119,44 +120,116 @@ def _stock_account_prepare_anglo_saxon_in_lines_vals(self):
...
-                        vals.update(line._get_fields_onchange_subtotal(price_subtotal=vals['price_subtotal']))
-                        lines_vals_list.append(vals)
+                    if move.currency_id.compare_amounts(price_unit, valuation_price_unit) != 0:
+                        accounts = line.product_id.product_tmpl_id.get_product_accounts()
+                        stock_valuation_account = accounts.get('stock_valuation')
+
+                        # Retrieve stock valuation moves.
+                        valuation_stock_moves = self.env['stock.move'].search([
+                            ('purchase_line_id', '=', line.purchase_line_id.id),
+                            ('product_qty', '!=', 0.0),
```
Comment (sle-odoo) [inline addons/purchase_stock/models/account_invoice.py]: why is it different than the valuation moves set at line 60 ?

File: addons/purchase_stock/models/account_invoice.py [hunk 5]
```diff
@@ -119,44 +120,116 @@ def _stock_account_prepare_anglo_saxon_in_lines_vals(self):
...
+                                'account_move_id': move.id,
+                                'stock_valuation_layer_id': linked_layer.id
+                            })
+                            # Update the standard price if product cost method is AVCO.
+                            if product.cost_method == 'average' and not float_is_zero(product.quantity_svl, precision_rounding=product.uom_id.rounding):
+                                product.with_company(self.company_id).sudo().with_context(disable_auto_svl=True).standard_price += price_unit_val_dif / product.quantity_svl
+
+                            # If the product already left the stock, create a
+                            # negative counter part for the stock valuation layer.
+                            product_already_out = linked_layer.remaining_qty == 0
```
Comment (sle-odoo) [inline addons/purchase_stock/models/account_invoice.py]: float_is_zero

File: addons/purchase_stock/models/account_invoice.py [hunk 6]
```diff
@@ -119,44 +120,116 @@ def _stock_account_prepare_anglo_saxon_in_lines_vals(self):
...
+                                    'description': _(svl_description + ' as the product is already out') % (move.purchase_id.name, move._get_next_sequence()),
+                                    'value': -price_subtotal,
+                                })
+                                svl_vals.append(svl_out_vals)
+                            else:
+                                linked_layer.remaining_value += price_subtotal
+
+                            self.env['stock.valuation.layer'].sudo().create(svl_vals)
+
+                            # Add price difference account line.
```
Comment (sle-odoo) [inline addons/purchase_stock/models/account_invoice.py]: this isn't a price difference account line

File: addons/purchase_stock/models/account_invoice.py [hunk 7]
```diff
@@ -119,44 +120,116 @@ def _stock_account_prepare_anglo_saxon_in_lines_vals(self):
...
+                                'name': line.name[:64],
+                                'move_id': move.id,
+                                'currency_id': line.currency_id.id,
+                                'product_id': product.id,
+                                'product_uom_id': line.product_uom_id.id,
+                                'quantity': line.quantity,
+                                'analytic_account_id': line.analytic_account_id.id,
+                                'analytic_tag_ids': [(6, 0, line.analytic_tag_ids.ids)],
+                                'exclude_from_invoice_tab': True,
+                                'is_anglo_saxon_line': True,
```
Comment (sle-odoo) [inline addons/purchase_stock/models/account_invoice.py]: i'm not sure this should be an anglo saxon line? is it intended?

**Comment by svs-odoo** [comment]
Current version of this PR: https://github.com/odoo/odoo/pull/89886

**Pull request closed without merging**